File: app/src/main/java/com/o4x/musical/ui/fragments/player/MiniPlayerFragment.java
package com.o4x.musical.ui.fragments.player;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;

import com.o4x.musical.R;
import com.o4x.musical.helper.MusicPlayerRemote;
import com.o4x.musical.helper.MusicProgressViewUpdateHelper;
import com.o4x.musical.helper.PlayPauseButtonOnClickHandler;
import com.o4x.musical.ui.fragments.AbsMusicServiceFragment;
import com.o4x.musical.util.color.MediaNotificationProcessor;
import com.o4x.musical.views.IconImageView;
import com.o4x.musical.drawables.PlayPauseDrawable;

import org.jetbrains.annotations.NotNull;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import code.name.monkey.appthemehelper.extensions.ColorExtKt;
import code.name.monkey.appthemehelper.util.ATHUtil;
import code.name.monkey.appthemehelper.util.ColorUtil;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;

/**
 * @author Karim Abou Zeid (kabouzeid)
 */
public class MiniPlayerFragment extends AbsMusicServiceFragment implements MusicProgressViewUpdateHelper.Callback {

    public MiniPlayerFragment() {
        super(R.layout.fragment_mini_player);
    }

    private Unbinder unbinder;

    @BindView(R.id.container)
    FrameLayout container;
    @BindView(R.id.mini_player_image)
    IconImageView miniPlayerImage;
    @BindView(R.id.mini_player_title)
    TextView miniPlayerTitle;
    @BindView(R.id.mini_player_play_pause_button)
    ImageView miniPlayerPlayPauseButton;
    @BindView(R.id.progress_bar_container)
    FrameLayout progressBarContainer;
    @BindView(R.id.progress_bar)
    MaterialProgressBar progressBar;

    private PlayPauseDrawable miniPlayerPlayPauseDrawable;

    private MusicProgressViewUpdateHelper progressViewUpdateHelper;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        progressViewUpdateHelper = new MusicProgressViewUpdateHelper(this);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);

        view.setOnTouchListener(new FlingPlayBackController(getActivity()));
        setUpMiniPlayer();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    private void setUpMiniPlayer() {
        setUpPlayPauseButton();
        setProgressColor(ColorExtKt.accentColor(this));
    }

    private void setProgressColor(@ColorInt int color) {
        progressBarContainer.setBackgroundColor(ColorUtil.INSTANCE.withAlpha(color, .3f));
        progressBar.setSupportProgressTintList(ColorStateList.valueOf(color));
    }

    private void setUpPlayPauseButton() {
        miniPlayerPlayPauseDrawable = new PlayPauseDrawable(getActivity());
        miniPlayerPlayPauseButton.setImageDrawable(miniPlayerPlayPauseDrawable);
        miniPlayerPlayPauseButton.setColorFilter(ATHUtil.INSTANCE.resolveColor(getActivity(), R.attr.iconColor, ColorExtKt.textColorSecondary(getActivity())), PorterDuff.Mode.SRC_IN);
        miniPlayerPlayPauseButton.setOnClickListener(new PlayPauseButtonOnClickHandler());
    }

    private void updateSongTitle() {
        miniPlayerTitle.setText(MusicPlayerRemote.getCurrentSong().getTitle());
    }

    @Override
    public void onServiceConnected() {
        updateSongTitle();
        updatePlayPauseDrawableState(false);
        onUpdateProgressViews(MusicPlayerRemote.getSongProgressMillis(),
                MusicPlayerRemote.getSongDurationMillis());
    }

    @Override
    public void onPlayingMetaChanged() {
        updateSongTitle();
    }

    @Override
    public void onPlayStateChanged() {
        updatePlayPauseDrawableState(true);
    }

    @Override
    public void onUpdateProgressViews(int progress, int total) {
        progressBar.setMax(total);
        progressBar.setProgress(progress);
    }

    @Override
    public void onResume() {
        super.onResume();
        progressViewUpdateHelper.start();
    }

    @Override
    public void onPause() {
        super.onPause();
        progressViewUpdateHelper.stop();
    }

    private static class FlingPlayBackController implements View.OnTouchListener {

        GestureDetector flingPlayBackController;

        public FlingPlayBackController(Context context) {
            flingPlayBackController = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
                @Override
                public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
                    if (Math.abs(velocityX) > Math.abs(velocityY)) {
                        if (velocityX < 0) {
                            MusicPlayerRemote.playNextSong();
                            return true;
                        } else if (velocityX > 0) {
                            MusicPlayerRemote.playPreviousSong();
                            return true;
                        }
                    }
                    return false;
                }
            });
        }

        @Override
        public boolean onTouch(View v, MotionEvent event) {
            return flingPlayBackController.onTouchEvent(event);
        }
    }

    protected void updatePlayPauseDrawableState(boolean animate) {
        if (MusicPlayerRemote.isPlaying()) {
            miniPlayerPlayPauseDrawable.setPause(animate);
        } else {
            miniPlayerPlayPauseDrawable.setPlay(animate);
        }
    }

    public void setColor(@NotNull MediaNotificationProcessor colors) {
        final int fg = colors.getPrimaryTextColor();

        container.setBackgroundColor(colors.getBackgroundColor());
        setProgressColor(fg);
        miniPlayerImage.setColorFilter(fg);
        miniPlayerPlayPauseButton.setColorFilter(fg, PorterDuff.Mode.SRC_IN);
        miniPlayerTitle.setTextColor(fg);

        // I want title slide show just when it colored
        miniPlayerTitle.setSelected(true);
    }
}

File: app/src/main/java/com/o4x/musical/ui/fragments/player/MiniPlayerFragment.kt
package com.o4x.musical.ui.fragments.player

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.PorterDuff
import android.os.Bundle
import android.view.*
import androidx.annotation.ColorInt
import code.name.monkey.appthemehelper.extensions.accentColor
import code.name.monkey.appthemehelper.extensions.textColorSecondary
import code.name.monkey.appthemehelper.util.ATHUtil.resolveColor
import code.name.monkey.appthemehelper.util.ColorUtil.withAlpha
import com.o4x.musical.R
import com.o4x.musical.ads.AdsUtils
import com.o4x.musical.databinding.FragmentMiniPlayerBinding
import com.o4x.musical.drawables.PlayPauseDrawable
import com.o4x.musical.helper.MusicPlayerRemote.currentSong
import com.o4x.musical.helper.MusicPlayerRemote.isPlaying
import com.o4x.musical.helper.MusicPlayerRemote.playNextSong
import com.o4x.musical.helper.MusicPlayerRemote.playPreviousSong
import com.o4x.musical.helper.PlayPauseButtonOnClickHandler
import com.o4x.musical.ui.fragments.AbsMusicServiceFragment
import com.o4x.musical.util.color.MediaNotificationProcessor
import kotlin.math.abs

/**
 * @author Karim Abou Zeid (kabouzeid)
 */
open class MiniPlayerFragment : AbsMusicServiceFragment(R.layout.fragment_mini_player) {

    private var _binding: FragmentMiniPlayerBinding? = null
    private val binding get() = _binding!!

    private var miniPlayerPlayPauseDrawable: PlayPauseDrawable? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentMiniPlayerBinding.inflate(inflater, container, false)
        binding.progressViewModel = serviceActivity.playerViewModel
        binding.lifecycleOwner = viewLifecycleOwner
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.setOnTouchListener(FlingPlayBackController(serviceActivity))
        setUpMiniPlayer()

        AdsUtils(requireActivity()).loadStandardBanner(binding.banner)
    }

    private fun setUpMiniPlayer() {
        setUpPlayPauseButton()
        setProgressColor(this.accentColor())
    }

    private fun setProgressColor(@ColorInt color: Int) {
        binding.progressBarContainer.setBackgroundColor(withAlpha(color, .3f))
        binding.progressBar.supportProgressTintList = ColorStateList.valueOf(color)
    }

    private fun setUpPlayPauseButton() {
        miniPlayerPlayPauseDrawable = PlayPauseDrawable(requireActivity())
        binding.miniPlayerPlayPauseButton.setImageDrawable(miniPlayerPlayPauseDrawable)
        binding.miniPlayerPlayPauseButton.setColorFilter(
            resolveColor(
                requireActivity(),
                R.attr.iconColor,
                requireActivity().textColorSecondary()
            ), PorterDuff.Mode.SRC_IN
        )
        binding.miniPlayerPlayPauseButton.setOnClickListener(PlayPauseButtonOnClickHandler())
    }

    private fun updateSongTitle() {
        binding.miniPlayerTitle.text = currentSong.title
    }

    override fun onServiceConnected() {
        updateSongTitle()
        updatePlayPauseDrawableState(false)
    }

    override fun onPlayingMetaChanged() {
        updateSongTitle()
    }

    override fun onMediaStoreChanged() {
        updateSongTitle()
    }

    override fun onPlayStateChanged() {
        updatePlayPauseDrawableState(true)
    }

    private class FlingPlayBackController(context: Context?) : View.OnTouchListener {

        var flingPlayBackController: GestureDetector

        @SuppressLint("ClickableViewAccessibility")
        override fun onTouch(v: View, event: MotionEvent): Boolean {
            return flingPlayBackController.onTouchEvent(event)
        }

        init {
            flingPlayBackController =
                GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
                    override fun onFling(
                        e1: MotionEvent,
                        e2: MotionEvent,
                        velocityX: Float,
                        velocityY: Float
                    ): Boolean {
                        if (abs(velocityX) > abs(velocityY)) {
                            if (velocityX < 0) {
                                playNextSong()
                                return true
                            } else if (velocityX > 0) {
                                playPreviousSong()
                                return true
                            }
                        }
                        return false
                    }
                })
        }
    }

    private fun updatePlayPauseDrawableState(animate: Boolean) {
        if (isPlaying) {
            miniPlayerPlayPauseDrawable!!.setPause(animate)
        } else {
            miniPlayerPlayPauseDrawable!!.setPlay(animate)
        }
    }

    fun setColor(colors: MediaNotificationProcessor) {
        val fg = colors.primaryTextColor
        binding.container.setBackgroundColor(colors.backgroundColor)
        setProgressColor(fg)
        binding.miniPlayerImage.setColorFilter(fg)
        binding.miniPlayerPlayPauseButton.setColorFilter(fg, PorterDuff.Mode.SRC_IN)
        binding.miniPlayerTitle.setTextColor(fg)

        // I want title slide show just when it colored
        binding.miniPlayerTitle.isSelected = true
    }
}